src/centralstore/incremental/NodeManager.cpp

```diff
@@ -0,0 +1,254 @@
+            }
+            if (!index_db.read(reinterpret_cast<char *>(&nodeIndexId), sizeof(unsigned int))) {
+                node_manager_logger.error("Error while reading index ID data from block i = " + std::to_string(i));
+            }
+            _nodeIndex[std::string(nodeIDC)] = nodeIndexId;
+        }
+    } else {
+        node_manager_logger.error("Error while opening the node index DB");
+    }
+
+    index_db.close();
+    return _nodeIndex;
+}
+
+unsigned int NodeManager::addRelation(NodeBlock source, NodeBlock destination) {
+    RelationBlock *newRelation = NULL;
+    if (source.edgeRef == 0 || destination.edgeRef == 0 ||
+        !source.searchRelation(destination)) {  // certainly a new relation block needed
+        newRelation = RelationBlock::add(source, destination);
+        if (newRelation) {
+            source.updateRelation(newRelation);
+            destination.updateRelation(newRelation);
+        } else {
+            node_manager_logger.error("Something went wrong while adding the new edge/relation source = " +
+                                      std::string(source.id) + " destination = " + std::string(destination.id));
+        }
+
+    } else {
+        node_manager_logger.warn("Relation/Edge already exist for source = " + std::string(source.id) +
+                                 " destination = " + std::string(destination.id));
+    }
+    return 0;
+}
+
+NodeBlock *NodeManager::addNode(std::string nodeId) {
+    unsigned int assignedNodeIndex;
+    if (this->nodeIndex.find(nodeId) == this->nodeIndex.end()) {
+        node_manager_logger.debug("NodeId not found in index for node ID " + nodeId);
+        NodeBlock *sourceBlk = new NodeBlock(nodeId, this->nextNodeIndex * NodeBlock::BLOCK_SIZE);
+        this->nodeIndex.insert({nodeId, this->nextNodeIndex});
+        assignedNodeIndex = this->nextNodeIndex;
+        this->nextNodeIndex++;
+        sourceBlk->save();
+        return sourceBlk;
+    } else {
+        node_manager_logger.debug("NodeId found in index for node ID " + nodeId);
+        return this->get(nodeId);
+    }
+}
+
+std::pair<NodeBlock, NodeBlock> NodeManager::addEdge(std::pair<std::string, std::string> edge) {
+    NodeBlock *sourceNode = this->addNode(edge.first);
+    NodeBlock *destNode = this->addNode(edge.second);
+    unsigned int relationAddr = this->addRelation(*sourceNode, *destNode);
+
+    node_manager_logger.debug("DEBUG: Source DB block address " + std::to_string(sourceNode->addr) +
+                              " Destination DB block address " + std::to_string(destNode->addr));
+    std::pair<NodeBlock, NodeBlock> returnVal = {*sourceNode, *destNode};
+    delete sourceNode; // to prevent memory leaks
+    delete destNode;
+    return returnVal;
+}
+
+int NodeManager::dbSize(std::string path) {
+    /*
+        The structure stat contains at least the following members:
+        st_dev     ID of device containing file
```

Comment:
change `ID of device` to `ID of the device`

Reply:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

Reply:
Done